src/repr/src/relation.rs

```diff
@@ -657,13 +819,285 @@ impl fmt::Display for NotNullViolation {
+            typ_idx: next_idx,
+            added: new_version,
+            dropped: None,
+        };
+
+        self.inner.typ.column_types.push(typ.into());
+        let prev = self.inner.metadata.insert(ColumnIndex(next_idx), col_meta);
+
+        assert_none!(prev, "column index overlap!");
+        self.validate();
+
+        new_version
+    }
+
+    /// Drops the column `name` from this [`RelationDesc`]. If there are multiple columns with
+    /// `name` drops the left-most one.
+    ///
+    /// # Panics
+    ///
+    /// Panics if a column with `name` does not exist or if the column was already dropped.
+    #[must_use]
+    pub fn drop_column<N>(&mut self, name: N) -> RelationVersion
+    where
+        N: Into<ColumnName>,
+    {
+        let name = name.into();
+        let latest_version = self.latest_version();
+        let new_version = latest_version.bump();
+
+        let col = self
+            .inner
+            .metadata
+            .values_mut()
+            .find(|meta| meta.name == name)
+            .expect("column to exist");
+
+        assert_none!(col.dropped, "column was already dropped");
+        col.dropped = Some(new_version);
+
+        self.validate();
+        new_version
+    }
+
+    /// Returns the [`RelationDesc`] at the latest version.
+    pub fn latest(&self) -> RelationDesc {
+        self.inner.clone()
+    }
+
+    /// Returns this [`RelationDesc`] at the specified version.
+    pub fn at_version(&self, version: RelationVersionSelector) -> RelationDesc {
+        // Get all of the changes from the start, up to whatever version was requested.
+        //
+        // TODO(parkmycar): We should probably panic on unknown verisons?
+        let up_to_version = match version {
+            RelationVersionSelector::Latest => RelationVersion(u64::MAX),
+            RelationVersionSelector::Specific(v) => v,
+        };
+
+        let valid_columns = self.inner.metadata.iter().filter(|(_col_idx, meta)| {
+            let added = meta.added <= up_to_version;
+            let dropped = meta
+                .dropped
+                .map(|dropped_at| up_to_version >= dropped_at)
+                .unwrap_or(false);
+
+            added && !dropped
+        });
+
+        let mut column_types = Vec::new();
+        let mut column_metas = BTreeMap::new();
+
+        // N.B. At this point we need to be careful because col_idx might not
+        // equal typ_idx.
+        for (col_idx, meta) in valid_columns {
+            let new_meta = ColumnMetadata {
+                name: meta.name.clone(),
+                typ_idx: column_types.len(),
+                added: meta.added.clone(),
+                dropped: meta.dropped.clone(),
+            };
+            column_types.push(self.inner.typ.columns()[meta.typ_idx].clone());
+            column_metas.insert(*col_idx, new_meta);
+        }
+
+        let relation_type = RelationType {
+            column_types,
+            keys: self.inner.typ.keys.clone(),
```

Comment:
What happens if `keys` reference a dropped column or a column that was added after `version`?

Reply:
Good callout! I don't think we'll allow dropping a column that is a key, at least not v0, so I made that panic. I didn't consider the case of keys getting re-ordered though so thank you for calling that out, I fixed this and added a test case